src/CaptureFile/CaptureFileOutputStream.cpp

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "CaptureFile/CaptureFileOutputStream.h"
+
+#include <absl/base/casts.h>
+#include <google/protobuf/io/coded_stream.h>
+#include <google/protobuf/io/zero_copy_stream_impl.h>
+
+#include <optional>
+#include <string>
+
+#include "CaptureFileConstants.h"
+#include "OrbitBase/File.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/SafeStrerror.h"
+
+namespace orbit_capture_file {
+
+namespace {
+
+class CaptureFileOutputStreamImpl final : public CaptureFileOutputStream {
+ public:
+  explicit CaptureFileOutputStreamImpl(std::filesystem::path path) : path_{std::move(path)} {}
+  ~CaptureFileOutputStreamImpl() noexcept override;
+
+  ErrorMessageOr<void> Initialize();
+  ErrorMessageOr<void> WriteCaptureEvent(
+      const orbit_grpc_protos::ClientCaptureEvent& event) override;
+  void Close() noexcept override;
+
+ private:
+  ErrorMessageOr<void> WriteHeader();
+  // Handles write error by cleaning up the file and generating error message.
+  ErrorMessage HandleWriteError(const char* section_name, std::string_view original_error);
+  // Call this in case of unrecoverable error to close and remove the file.
+  void CloseAndTryRemoveFile();
```

Comment:
Should we get notified, when it was not able to remove?

Reply:
Make sense. Thanks

Reply:
Maybe rename to `CloseAndTryRemoveFileAfterError` then?

Reply:
done.